Rolling back a failing transaction through the HTTPRepositoryConnection results in transaction not found

If for some reason the transaction failed inside the database (sail implementation side), there is no way to close or rollback the transaction properly by the client.

Example code on the client side:
```java
try {
    HTTPRepositoryConnection connection = ....
    connection.add(statement); // something that will cause a transaction failure
    connection.commit();
} catch (Exception ex) {
    connection.rollback(); // this causes the missing transaction message
} finally {
    connection.close(); // this also causes it because the transaction is active and the connection wants to rollback.
}
```

Here is a full stacktrace on the server
```
[ERROR] 2017-01-17 18:02:49,783 [repositories/graphdb_standard_test_1_1484668930169 | o.e.r.h.s.ProtocolExceptionResolver] Error while handling request
org.eclipse.rdf4j.repository.RepositoryException: transaction with id 08f1e560-4308-402c-9414-62d64e158d0e not registered.
	at org.eclipse.rdf4j.http.server.repository.transaction.ActiveTransactionRegistry.getTransaction(ActiveTransactionRegistry.java:139)
	at org.eclipse.rdf4j.http.server.repository.transaction.TransactionController.handleRequestInternal(TransactionController.java:113)
	at org.springframework.web.servlet.mvc.AbstractController.handleRequest(AbstractController.java:146)
	at org.springframework.web.servlet.mvc.SimpleControllerHandlerAdapter.handle(SimpleControllerHandlerAdapter.java:50)
	at org.springframework.web.servlet.DispatcherServlet.doDispatch(DispatcherServlet.java:959)
	at org.springframework.web.servlet.DispatcherServlet.doService(DispatcherServlet.java:893)
	at org.springframework.web.servlet.FrameworkServlet.processRequest(FrameworkServlet.java:966)
	at org.springframework.web.servlet.FrameworkServlet.doDelete(FrameworkServlet.java:890)
	at javax.servlet.http.HttpServlet.service(HttpServlet.java:654)
	at org.springframework.web.servlet.FrameworkServlet.service(FrameworkServlet.java:842)
	at javax.servlet.http.HttpServlet.service(HttpServlet.java:729)
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:292)
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:207)
	at com.github.ziplet.filter.compression.CompressingFilter.doFilter(CompressingFilter.java:300)
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:240)
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:207)
	at org.apache.catalina.core.StandardWrapperValve.invoke(StandardWrapperValve.java:212)
	at org.apache.catalina.core.StandardContextValve.invoke(StandardContextValve.java:106)
	at org.apache.catalina.authenticator.AuthenticatorBase.invoke(AuthenticatorBase.java:502)
	at org.apache.catalina.core.StandardHostValve.invoke(StandardHostValve.java:141)
	at org.apache.catalina.valves.ErrorReportValve.invoke(ErrorReportValve.java:79)
	at org.apache.catalina.core.StandardEngineValve.invoke(StandardEngineValve.java:88)
	at org.apache.catalina.connector.CoyoteAdapter.service(CoyoteAdapter.java:522)
	at org.apache.coyote.http11.AbstractHttp11Processor.process(AbstractHttp11Processor.java:1095)
	at org.apache.coyote.AbstractProtocol$AbstractConnectionHandler.process(AbstractProtocol.java:672)
	at org.apache.tomcat.util.net.NioEndpoint$SocketProcessor.doRun(NioEndpoint.java:1500)
	at org.apache.tomcat.util.net.NioEndpoint$SocketProcessor.run(NioEndpoint.java:1456)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
	at org.apache.tomcat.util.threads.TaskThread$WrappingRunnable.run(TaskThread.java:61)
	at java.lang.Thread.run(Thread.java:745)
```
and on the client side 
```
org.eclipse.rdf4j.repository.RepositoryException: org.eclipse.rdf4j.repository.RepositoryException: transaction with id 0e5cb052-0eef-4fc8-86e2-27f8d2c41ea9 not registered.

	at org.eclipse.rdf4j.repository.http.HTTPRepositoryConnection.rollback(HTTPRepositoryConnection.java:357)
	at org.eclipse.rdf4j.repository.http.HTTPRepositoryConnection.close(HTTPRepositoryConnection.java:373)
	at com.ontotext.test.functional.base.AbstractInconsistencyInference.testStatementWithChecks(AbstractInconsistencyInference.java:54)
	at com.ontotext.test.functional.base.AbstractInconsistencyInference.testWithAsymmetricProperty(AbstractInconsistencyInference.java:129)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:47)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:44)
	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
	at org.junit.rules.ExternalResource$1.evaluate(ExternalResource.java:48)
	at org.junit.rules.ExternalResource$1.evaluate(ExternalResource.java:48)
	at org.junit.rules.ExternalResource$1.evaluate(ExternalResource.java:48)
	at com.ontotext.test.Repeat$1.evaluate(Repeat.java:40)
	at org.junit.rules.RunRules.evaluate(RunRules.java:20)
	at org.junit.runners.ParentRunner.runLeaf(ParentRunner.java:271)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:70)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:50)
	at org.junit.runners.ParentRunner$3.run(ParentRunner.java:238)
	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:63)
	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:236)
	at org.junit.runners.ParentRunner.access$000(ParentRunner.java:53)
	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:229)
	at org.junit.runners.ParentRunner.run(ParentRunner.java:309)
	at org.junit.runner.JUnitCore.run(JUnitCore.java:160)
	at com.intellij.junit4.JUnit4IdeaTestRunner.startRunnerWithArgs(JUnit4IdeaTestRunner.java:68)
	at com.intellij.rt.execution.junit.IdeaTestRunner$Repeater.startRunnerWithArgs(IdeaTestRunner.java:51)
	at com.intellij.rt.execution.junit.JUnitStarter.prepareStreamsAndStart(JUnitStarter.java:237)
	at com.intellij.rt.execution.junit.JUnitStarter.main(JUnitStarter.java:70)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at com.intellij.rt.execution.application.AppMain.main(AppMain.java:147)
Caused by: org.eclipse.rdf4j.repository.RepositoryException: transaction with id 0e5cb052-0eef-4fc8-86e2-27f8d2c41ea9 not registered.
	at org.eclipse.rdf4j.http.client.SparqlSession.execute(SparqlSession.java:1132)
	at org.eclipse.rdf4j.http.client.SesameSession.rollbackTransaction(SesameSession.java:553)
	at org.eclipse.rdf4j.repository.http.HTTPRepositoryConnection.rollback(HTTPRepositoryConnection.java:353)
	... 35 more

```

this affects rdf4j 2.1.5 and is caused by https://github.com/eclipse/rdf4j/commit/5880ad6842ac493eac4efcf74bb295e992877b70 where the transaction is deregistered in a finally on commit. This is not correct because, if the transaction failed, it should stay in the registry.

--- a/core/http/server-spring/src/main/java/org/eclipse/rdf4j/http/server/repository/transaction/TransactionController.java
+++ b/core/http/server-spring/src/main/java/org/eclipse/rdf4j/http/server/repository/transaction/TransactionController.java
@@ -255,12 +255,8 @@ public class TransactionController extends AbstractController {
 				case UPDATE:
 					return getSparqlUpdateResult(transaction, request, response);
 				case COMMIT:
-					try {
-						transaction.commit();
-					}
-					finally {
-						ActiveTransactionRegistry.INSTANCE.deregister(transaction);
-					}
+					transaction.commit();
+					ActiveTransactionRegistry.INSTANCE.deregister(transaction);
 					break;
 				default:
 					logger.warn("transaction modification action '{}' not recognized", action);

--- a/core/http/server-spring/src/main/java/org/eclipse/rdf4j/http/server/repository/transaction/TransactionController.java
+++ b/core/http/server-spring/src/main/java/org/eclipse/rdf4j/http/server/repository/transaction/TransactionController.java
@@ -256,6 +256,8 @@ public class TransactionController extends AbstractController {
 					return getSparqlUpdateResult(transaction, request, response);
 				case COMMIT:
 					transaction.commit();
+					// If commit fails with an exception, deregister should be skipped so the user
+					// has a chance to do a proper rollback. See #725.
 					ActiveTransactionRegistry.INSTANCE.deregister(transaction);
 					break;
 				default:

